cocotb/cocotb.bzl

```diff
@@ -0,0 +1,267 @@
+    testcase_args = _list_to_argstring(ctx.attr.testcase, "testcase")
+    build_args = _list_to_argstring(ctx.attr.build_args, "build_args")
+    gpi_interfaces_args = _list_to_argstring(ctx.attr.gpi_interfaces, "gpi_interfaces")
+    test_args = _list_to_argstring(ctx.attr.test_args, "test_args")
+    plus_args = _list_to_argstring(ctx.attr.plus_args, "plus_args")
+    extra_env_args = _list_to_argstring(ctx.attr.extra_env, "extra_env")
+
+    defines_args = _dict_to_argstring(ctx.attr.defines, "defines")
+    parameters_args = _dict_to_argstring(ctx.attr.parameters, "parameters")
+
+    verbose_args = " --verbose" if ctx.attr.verbose else ""
+    seed_args = " --seed {}".format(ctx.attr.seed) if ctx.attr.seed != "" else ""
+
+    test_module_args = _pymodules_to_argstring(ctx.files.test_module, "test_module")
+
+    # define a script and a command
+    runner_script = ctx.actions.declare_file("cocotb_runner.sh")
+
+    sim_paths = _remove_duplicates_from_list([dep.label.workspace_root for dep in ctx.attr.sim])
+    path = ":".join(["$PWD/" + str(p) for p in sim_paths])
+
+    command = (
+        "PATH={}:$PATH ".format(path) +
+        "python {}".format(ctx.executable._cocotb_wrapper.short_path) +
+        " --sim {}".format(ctx.attr.sim_name) +
+        " --hdl_library {}".format(ctx.attr.hdl_library) +
+        " --hdl_toplevel {}".format(ctx.attr.hdl_toplevel) +
+        " --hdl_toplevel_lang {}".format(ctx.attr.hdl_toplevel_lang) +
+        verilog_sources_args +
+        vhdl_sources_args +
+        includes_args +
+        testcase_args +
+        build_args +
+        gpi_interfaces_args +
+        test_args +
+        plus_args +
+        extra_env_args +
+        defines_args +
+        parameters_args +
+        verbose_args +
+        seed_args +
+        test_module_args
+    )
+
+    ctx.actions.write(output = runner_script, content = command)
+
+    # specify dependencies for the script
+    py_toolchain = ctx.toolchains["@bazel_tools//tools/python:toolchain_type"].py3_runtime
+    transitive_files = depset(
+        direct = [py_toolchain.interpreter],
+        transitive = [dep[PyInfo].transitive_sources for dep in ctx.attr.deps] +
+                     [ctx.attr._cocotb_wrapper[PyInfo].transitive_sources] +
+                     [py_toolchain.files],
+    )
+
+    runfiles = ctx.runfiles(
+        files = ctx.files._cocotb_wrapper +
+                verilog_files +
+                vhdl_files +
+                ctx.files.test_module,
+        transitive_files = transitive_files,
+    ).merge_all(
+        [dep.default_runfiles for dep in ctx.attr.deps] +
+        [dep.default_runfiles for dep in ctx.attr.test_module] +
+        [dep.default_runfiles for dep in ctx.attr.sim],
+    )
+
+    # specify pythonpath for the script
+    test_module_paths = _remove_duplicates_from_list([module.dirname for module in ctx.files.test_module])
+    pypath = ":".join([str(p) for p in test_module_paths])
```

Comment:
I'm not a 100% sure on this, but I believe you're supposed to use the Bazel Python "Configuration Fragment" to get the pypath. Could definitely be wrong though.

https://bazel.build/extending/rules#configuration_fragments
https://bazel.build/rules/lib/fragments/bazel_py#python_path

Reply:
It seems that the `ctx.fragments.bazel_py.python_path` contains information about the `--python-path` flag which seems to have completely different meaning - the location of the python interpreter (see https://github.com/bazelbuild/bazel/blob/8715e9ae6f5c2fd42607018b3adaad14bf601a1c/src/main/starlark/builtins_bzl/common/python/py_executable.bzl#L217) and may be removed (see https://github.com/bazelbuild/bazel/issues/7901). The variable that I set gives the interpreter additional search paths for module files (see https://docs.python.org/3/using/cmdline.html#envvar-PYTHONPATH)